src/rgw/driver/rados/topic_migration.cc

```diff
@@ -0,0 +1,247 @@
+
+  auto bucket_name_marker = bucket_topics_oid.substr(pos + rgw::sal::pubsub_bucket_oid_infix.size());
+  pos = bucket_name_marker.find('/');
+  if (pos == std::string::npos) {
+    ldpp_dout(dpp, 1) << "ERROR: bucket_topics_oid:" << bucket_topics_oid << " doesn't contain / after bucket name!" << dendl;
+    return -EINVAL;
+  }
+  bucket_name = bucket_name_marker.substr(0, pos);
+  marker = bucket_name_marker.substr(pos + 1);
+
+  return 0;
+}
+
+// migrate v1 notification metadata for a single bucket
+int migrate_notification(const DoutPrefixProvider* dpp, optional_yield y,
+                         rgw::sal::RadosStore* driver, const rgw_raw_obj& obj)
+{
+  // parse bucket name and marker of out "pubsub.{tenant}.bucket.{name}/{marker}"
+  std::string tenant;
+  std::string bucket_name;
+  std::string marker;
+  int r = deconstruct_topics_oid(obj.oid, tenant, bucket_name, marker, dpp);
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to read tenant, bucket name and marker from " << obj
+                      << " with: " << cpp_strerror(r)
+                      << " expected format pubsub.{tenant}.bucket.{name}/{marker}!" << dendl;
+    return r;
+  }
+
+  // migrate the notifications
+  rgw_pubsub_bucket_topics bucket_topics;
+  rgw_bucket rgw_bucket_info(tenant, bucket_name);
+  rgw_bucket_info.marker = marker;
+  rgw::sal::RadosBucket rados_bucket(driver, rgw_bucket_info);
+  RGWObjVersionTracker bucket_topics_objv;
+  r = rados_bucket.read_topics(bucket_topics, &bucket_topics_objv, y, dpp);
+  if (r == -ENOENT) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "failed to read v1 bucket notifications from " << obj
+                      << " with: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  // in a for-loop that retries ECANCELED errors:
+  // {
+  // load the corresponding bucket by name
+  // break if marker doesn't match loaded bucket's
+  // break if RGW_ATTR_BUCKET_NOTIFICATION xattr already exists
+  // write RGW_ATTR_BUCKET_NOTIFICATION xattr
+  // }
+  std::unique_ptr<rgw::sal::Bucket> bucket;
+  do {
+    r = driver->load_bucket(dpp, rgw_bucket_info, &bucket, y);
+    if (r == -ENOENT) {
+      break; // bucket is deleted, we should delete the v1 notification
+    }
+    if (r < 0) {
+      ldpp_dout(dpp, 1) << "failed to load the bucket from " << obj
+                        << " with: " << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    if (bucket->get_marker() != marker || bucket->get_attrs().contains(RGW_ATTR_BUCKET_NOTIFICATION)) {
+      break;
+    }
+
+    bufferlist bl;
+    bucket_topics.encode(bl);
+    bucket->get_attrs()[RGW_ATTR_BUCKET_NOTIFICATION] = std::move(bl);
+    r = bucket->put_info(dpp, false, real_time(), y);
+    if (r != -ECANCELED && r < 0) {
+      ldpp_dout(dpp, 1) << "ERROR: failed writing bucket instance info: " << cpp_strerror(-r) << dendl;
+      return r;
+    }
+  } while (r != -ECANCELED);
```

Comment:
we probably want to limit the number of times we'll retry here. you can see `retry_raced_bucket_write()` in rgw_op.cc for an example